lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -123,76 +253,53 @@ combineLsqData NodePoolLsqData{nOpt, rewards, stake} =
-    }
-
-newStakePoolLayer
-    :: GenesisParameters
-    -> NetworkLayer IO (IO Shelley) b
-    -> StakePoolLayer
-newStakePoolLayer gp nl = StakePoolLayer
-    { knownPools = _knownPools
-    , listStakePools = _listPools
-    }
+-- | Combines all the chain-following data into a single map
+-- (doesn't include metadata)
+combineChainData
+    :: Map PoolId PoolRegistrationCertificate
+    -> Map PoolId (Quantity "block" Word64)
+    -> Map PoolId
+        (PoolRegistrationCertificate, Quantity "block" Word64)
+combineChainData =
+    Map.merge registeredNoProductions notRegisteredButProducing bothPresent
   where
-    dummyCoin = Coin 0
+    registeredNoProductions  = traverseMissing $ \_k cert ->
+        pure (cert, Quantity 0)
 
-    -- Note: We shouldn't have to do this conversion.
-    el = getEpochLength gp
-    gh = getGenesisBlockHash gp
-    getTip = fmap (toPoint gh el) . liftIO $ unsafeRunExceptT $ currentNodeTip nl
+    -- Ignore blocks produced by BFT nodes.
+    notRegisteredButProducing = dropMissing
 
-    _knownPools
-        :: IO [PoolId]
-    _knownPools = do
-        pt <- getTip
-        res <- runExceptT $ map fst . Map.toList
-            . combineLsqData <$> stakeDistribution nl pt dummyCoin
-        case res of
-            Right x -> return x
-            Left _e -> return []
-
-
-    _listPools
-        :: Coin
-        -- ^ The amount of stake the user intends to delegate, which may affect the
-        -- ranking of the pools.
-        -> ExceptT ErrNetworkUnavailable IO [Api.ApiStakePool]
-    _listPools s = do
-            pt <- liftIO getTip
-            map mkApiPool
-                . sortOn (Down . nonMyopicMemberRewards . snd)
-                . Map.toList
-                . combineLsqData
-                <$> stakeDistribution nl pt s
-      where
-        mkApiPool (pid, PoolLsqMetrics prew pstk psat) = Api.ApiStakePool
-            { Api.id = (ApiT pid)
-            , Api.metrics = Api.ApiStakePoolMetrics
-                { Api.nonMyopicMemberRewards = (mapQ fromIntegral prew)
-                , Api.relativeStake = Quantity pstk
-                , Api.saturation = psat
-                , Api.producedBlocks = Quantity 0 -- TODO: Implement
-                }
-            , Api.metadata = Nothing -- TODO: Implement
-            , Api.cost = Quantity 0 -- TODO: Implement
-            , Api.margin = Quantity $ unsafeMkPercentage 0 -- TODO: Implement
-            }
+    bothPresent = zipWithMatched $ const (,)
 
-        mapQ f (Quantity x) = Quantity $ f x
+-- NOTE: If performance becomes a problem, we could try replacing all
+-- the individual DB queries, and combbination functions with a single
+-- hand-written Sqlite query.
+readDBPoolData
+    :: DBLayer IO
+    -> IO (Map PoolId PoolDBMetrics)
+readDBPoolData DBLayer{..} = atomically $ do
+    pools <- listRegisteredPools
+    certMap <- Map.fromList . catMaybes <$> forM pools
+        (\poolId -> ((fmap (poolId,)) <$> readPoolRegistration poolId))
```

Comment:
A bit simpler...

```suggestion
    registrations <- mapM readPoolRegistration pools
    let certMap = Map.fromList [(poolId, cert) | (poolId, Just cert) <- zip pools registrations]
```

Reply:
`zip` feels a bit wrong to me, but sure, it does look neat